// angular-file/esm5/file-upload/ngfFormData.directive.js
import * as tslib_1 from "tslib";
import { IterableDiffer, IterableDiffers, Directive, EventEmitter, Output, Input } from '@angular/core';
var ngfFormData = /** @class */ (function () {
    function ngfFormData(IterableDiffers) {
        this.postName = "file";
        this.FormData = new FormData();
        this.FormDataChange = new EventEmitter();
        this.differ = IterableDiffers.find([]).create();
    }
    ngfFormData.prototype.ngDoCheck = function () {
        var _this = this;
        var changes = this.differ.diff(this.files);
        if (changes) {
            setTimeout(function () { return _this.buildFormData(); }, 0);
        }
    };
    ngfFormData.prototype.buildFormData = function () {
        var _this = this;
        var isArray = typeof (this.files) === 'object' && this.files.constructor === Array;
        if (isArray) {
            this.FormData = new FormData();
            var files = this.files || [];
            files.forEach(function (file) {
                return _this.FormData.append(_this.postName, file, _this.fileName || file.name);
            });
            this.FormDataChange.emit(this.FormData);
        }
        else {
            delete this.FormData;
        }
    };
    ngfFormData.ctorParameters = function () { return [
        { type: IterableDiffers }
    ]; };
    tslib_1.__decorate([
        Input()
    ], ngfFormData.prototype, "files", void 0);
    tslib_1.__decorate([
        Input()
    ], ngfFormData.prototype, "postName", void 0);
    tslib_1.__decorate([
        Input()
    ], ngfFormData.prototype, "fileName", void 0);
    tslib_1.__decorate([
        Input()
    ], ngfFormData.prototype, "FormData", void 0);
    tslib_1.__decorate([
        Output()
    ], ngfFormData.prototype, "FormDataChange", void 0);
    ngfFormData = tslib_1.__decorate([
        Directive({ selector: 'ngfFormData' })
    ], ngfFormData);
    return ngfFormData;
}());
export { ngfFormData };
