// ngx-google-analytics/esm5/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of ngx-google-analytics
 */
export { GoogleAnalyticsInitializer, NGX_GOOGLE_ANALYTICS_INITIALIZER_PROVIDER } from './lib/initializers/google-analytics.initializer';
export { GoogleAnalyticsRouterInitializer, NGX_GOOGLE_ANALYTICS_ROUTER_INITIALIZER_PROVIDER } from './lib/initializers/google-analytics-router.initializer';
export {} from './lib/interfaces/i-google-analytics-commant';
export {} from './lib/interfaces/i-google-analytics-settings';
export { NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN } from './lib/tokens/ngx-google-analytics-settings-token';
export {} from './lib/types/ga-action.type';
export {} from './lib/types/ga-bind.type';
export { GoogleAnalyticsService } from './lib/services/google-analytics.service';
export { GaEventCategoryDirective } from './lib/directives/ga-event-category.directive';
export { GaEventDirective } from './lib/directives/ga-event.directive';
export { GaEventFormInputDirective } from './lib/directives/ga-event-form-input.directive';
export { NgxGoogleAnalyticsModule } from './lib/ngx-google-analytics.module';
export { NgxGoogleAnalyticsRouterModule } from './lib/ngx-google-analytics-router/ngx-google-analytics-router.module';
