// ngx-google-analytics/esm2015/lib/initializers/google-analytics-router.initializer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { APP_BOOTSTRAP_LISTENER } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';
import { GoogleAnalyticsService } from '../services/google-analytics.service';
/** @type {?} */
export const NGX_GOOGLE_ANALYTICS_ROUTER_INITIALIZER_PROVIDER = {
    provide: APP_BOOTSTRAP_LISTENER,
    multi: true,
    useFactory: GoogleAnalyticsRouterInitializer,
    deps: [
        GoogleAnalyticsService,
        Router
    ]
};
/**
 * @param {?} $gaService
 * @param {?} $router
 * @return {?}
 */
export function GoogleAnalyticsRouterInitializer($gaService, $router) {
    return (/**
     * @param {?} c
     * @return {?}
     */
    (c) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        $router
            .events
            .subscribe((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            if (event instanceof NavigationEnd) {
                $gaService.pageView(event.urlAfterRedirects, undefined, document.location.href);
            }
        }));
    }));
}
