// ngx-google-analytics/esm5/lib/initializers/google-analytics-router.initializer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { APP_BOOTSTRAP_LISTENER } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';
import { GoogleAnalyticsService } from '../services/google-analytics.service';
/** @type {?} */
export var NGX_GOOGLE_ANALYTICS_ROUTER_INITIALIZER_PROVIDER = {
    provide: APP_BOOTSTRAP_LISTENER,
    multi: true,
    useFactory: GoogleAnalyticsRouterInitializer,
    deps: [
        GoogleAnalyticsService,
        Router
    ]
};
/**
 * @param {?} $gaService
 * @param {?} $router
 * @return {?}
 */
export function GoogleAnalyticsRouterInitializer($gaService, $router) {
    var _this = this;
    return (/**
     * @param {?} c
     * @return {?}
     */
    function (c) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            $router
                .events
                .subscribe((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event instanceof NavigationEnd) {
                    $gaService.pageView(event.urlAfterRedirects, undefined, document.location.href);
                }
            }));
            return [2 /*return*/];
        });
    }); });
}
