// ngx-google-analytics/esm5/lib/initializers/google-analytics.initializer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { APP_INITIALIZER, isDevMode } from '@angular/core';
import { NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN } from '../tokens/ngx-google-analytics-settings-token';
/** @type {?} */
export var NGX_GOOGLE_ANALYTICS_INITIALIZER_PROVIDER = {
    provide: APP_INITIALIZER,
    multi: true,
    useFactory: GoogleAnalyticsInitializer,
    deps: [
        NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN
    ]
};
/**
 * @param {?} $settings
 * @return {?}
 */
export function GoogleAnalyticsInitializer($settings) {
    var _this = this;
    return (/**
     * @return {?}
     */
    function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var e_1, _a, initialCommands, _b, _c, command, s, head;
        return tslib_1.__generator(this, function (_d) {
            if (!$settings.trackingCode) {
                if (!isDevMode()) {
                    console.error('Empty tracking code for Google Analytics. Make sure to provide one when initializing NgxGoogleAnalyticsModule.');
                }
                return [2 /*return*/];
            }
            // Set default ga.js uri
            $settings.uri = $settings.uri || "https://www.googletagmanager.com/gtag/js?id=" + $settings.trackingCode;
            // these commands should run first!
            initialCommands = [
                { command: 'js', values: [new Date()] },
                { command: 'config', values: [$settings.trackingCode] }
            ];
            $settings.initCommands = tslib_1.__spread(initialCommands, ($settings.initCommands || []));
            window['dataLayer'] = window['dataLayer'] || [];
            window['gtag'] = window['gtag'] || (/**
             * @return {?}
             */
            function () {
                window['dataLayer'].push(arguments);
            });
            try {
                for (_b = tslib_1.__values($settings.initCommands), _c = _b.next(); !_c.done; _c = _b.next()) {
                    command = _c.value;
                    window['gtag'].apply(window, tslib_1.__spread([command.command], command.values));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            s = document.createElement('script');
            s.async = true;
            s.src = $settings.uri;
            head = document.getElementsByTagName('head')[0];
            head.appendChild(s);
            return [2 /*return*/];
        });
    }); });
}
