// angular-file/esm2015/file-upload/ngfDrop.directive.js
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, HostListener, Input, Output } from '@angular/core';
import { ngf } from "./ngf.directive";
let ngfDrop = class ngfDrop extends ngf {
    constructor() {
        super(...arguments);
        this.fileOver = new EventEmitter();
        this.validDrag = false;
        this.validDragChange = new EventEmitter();
        this.invalidDrag = false;
        this.invalidDragChange = new EventEmitter();
        this.dragFilesChange = new EventEmitter();
    }
    onDrop(event) {
        if (this.fileDropDisabled) {
            this.stopEvent(event);
            return;
        }
        this.closeDrags();
        let files = this.eventToFiles(event);
        if (!files.length)
            return;
        this.stopEvent(event);
        this.handleFiles(files);
    }
    handleFiles(files) {
        this.fileOver.emit(false); //turn-off dragover
        super.handleFiles(files);
    }
    onDragOver(event) {
        if (this.fileDropDisabled) {
            this.stopEvent(event);
            return;
        }
        const transfer = this.eventToTransfer(event);
        let files = this.eventToFiles(event);
        let jsonFiles = this.filesToWriteableObject(files);
        this.dragFilesChange.emit(this.dragFiles = jsonFiles);
        if (files.length) {
            this.validDrag = this.isFilesValid(files);
        }
        else {
            //Safari, IE11 & some browsers do NOT tell you about dragged files until dropped. Always consider a valid drag
            this.validDrag = true;
        }
        this.validDragChange.emit(this.validDrag);
        this.invalidDrag = !this.validDrag;
        this.invalidDragChange.emit(this.invalidDrag);
        transfer.dropEffect = 'copy'; //change cursor and such
        this.stopEvent(event);
        this.fileOver.emit(true);
    }
    closeDrags() {
        delete this.validDrag;
        this.validDragChange.emit(this.validDrag);
        this.invalidDrag = false;
        this.invalidDragChange.emit(this.invalidDrag);
        delete this.dragFiles;
        this.dragFilesChange.emit(this.dragFiles);
    }
    onDragLeave(event) {
        if (this.fileDropDisabled) {
            this.stopEvent(event);
            return;
        }
        this.closeDrags();
        if (this.element) {
            if (event.currentTarget === this.element[0]) {
                return;
            }
        }
        this.stopEvent(event);
        this.fileOver.emit(false);
    }
};
tslib_1.__decorate([
    Output()
], ngfDrop.prototype, "fileOver", void 0);
tslib_1.__decorate([
    Input()
], ngfDrop.prototype, "validDrag", void 0);
tslib_1.__decorate([
    Output()
], ngfDrop.prototype, "validDragChange", void 0);
tslib_1.__decorate([
    Input()
], ngfDrop.prototype, "invalidDrag", void 0);
tslib_1.__decorate([
    Output()
], ngfDrop.prototype, "invalidDragChange", void 0);
tslib_1.__decorate([
    Input()
], ngfDrop.prototype, "dragFiles", void 0);
tslib_1.__decorate([
    Output()
], ngfDrop.prototype, "dragFilesChange", void 0);
tslib_1.__decorate([
    HostListener('drop', ['$event'])
], ngfDrop.prototype, "onDrop", null);
tslib_1.__decorate([
    HostListener('dragover', ['$event'])
], ngfDrop.prototype, "onDragOver", null);
tslib_1.__decorate([
    HostListener('dragleave', ['$event'])
], ngfDrop.prototype, "onDragLeave", null);
ngfDrop = tslib_1.__decorate([
    Directive({
        selector: "[ngfDrop]",
        exportAs: "ngfDrop"
    })
], ngfDrop);
export { ngfDrop };
