// ngx-google-analytics/esm5/lib/ngx-google-analytics.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NGX_GOOGLE_ANALYTICS_INITIALIZER_PROVIDER } from './initializers/google-analytics.initializer';
import { NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN } from './tokens/ngx-google-analytics-settings-token';
import { GaEventDirective } from './directives/ga-event.directive';
import { GaEventCategoryDirective } from './directives/ga-event-category.directive';
import { GaEventFormInputDirective } from './directives/ga-event-form-input.directive';
var NgxGoogleAnalyticsModule = /** @class */ (function () {
    function NgxGoogleAnalyticsModule() {
    }
    /**
     * @param {?} trackingCode
     * @param {?=} commands
     * @param {?=} uri
     * @return {?}
     */
    NgxGoogleAnalyticsModule.forRoot = /**
     * @param {?} trackingCode
     * @param {?=} commands
     * @param {?=} uri
     * @return {?}
     */
    function (trackingCode, commands, uri) {
        if (commands === void 0) { commands = []; }
        return {
            ngModule: NgxGoogleAnalyticsModule,
            providers: [
                {
                    provide: NGX_GOOGLE_ANALYTICS_SETTINGS_TOKEN,
                    useValue: {
                        trackingCode: trackingCode,
                        commands: commands,
                        uri: uri
                    }
                },
                NGX_GOOGLE_ANALYTICS_INITIALIZER_PROVIDER,
            ]
        };
    };
    NgxGoogleAnalyticsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [GaEventDirective, GaEventCategoryDirective, GaEventFormInputDirective],
                    exports: [GaEventDirective, GaEventCategoryDirective, GaEventFormInputDirective]
                },] }
    ];
    return NgxGoogleAnalyticsModule;
}());
export { NgxGoogleAnalyticsModule };
