// ngx-google-analytics/esm2015/lib/directives/ga-event.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Optional, Input, HostListener } from '@angular/core';
import { GaEventCategoryDirective } from './ga-event-category.directive';
import { GoogleAnalyticsService } from '../services/google-analytics.service';
export class GaEventDirective {
    /**
     * @param {?} $gaCategoryDirective
     * @param {?} $gaService
     */
    constructor($gaCategoryDirective, $gaService) {
        this.$gaCategoryDirective = $gaCategoryDirective;
        this.$gaService = $gaService;
        this.gaBind = 'click';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    onClick() {
        if (this.gaBind === 'click') {
            this.trigger();
        }
    }
    /**
     * @return {?}
     */
    onFocus() {
        if (this.gaBind === 'focus') {
            this.trigger();
        }
    }
    /**
     * @return {?}
     */
    onBlur() {
        if (this.gaBind === 'blur') {
            this.trigger();
        }
    }
    /**
     * @protected
     * @return {?}
     */
    trigger() {
        try {
            if (!this.$gaCategoryDirective) {
                throw new Error('You must provide a gaCategory attribute w/ gaEvent Directive.');
            }
            if (!this.gaAction) {
                throw new Error('You must provide a gaAction atrribute to identify this event.');
            }
            this.$gaService
                .event(this.gaAction, this.$gaCategoryDirective.gaCategory, this.gaLabel || this.label, this.gaValue, this.gaInteraction);
        }
        catch (err) {
            this.throw(err);
        }
    }
    /**
     * @protected
     * @param {?} err
     * @return {?}
     */
    throw(err) {
        if (console && console.warn) {
            console.warn(err);
        }
    }
}
GaEventDirective.decorators = [
    { type: Directive, args: [{
                selector: `[gaEvent]`,
                exportAs: 'gaEvent'
            },] }
];
/** @nocollapse */
GaEventDirective.ctorParameters = () => [
    { type: GaEventCategoryDirective, decorators: [{ type: Optional }] },
    { type: GoogleAnalyticsService }
];
GaEventDirective.propDecorators = {
    gaAction: [{ type: Input }],
    gaLabel: [{ type: Input }],
    label: [{ type: Input }],
    gaValue: [{ type: Input }],
    gaInteraction: [{ type: Input }],
    gaBind: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }],
    onFocus: [{ type: HostListener, args: ['focus',] }],
    onBlur: [{ type: HostListener, args: ['blur',] }]
};
if (false) {
    /** @type {?} */
    GaEventDirective.prototype.gaAction;
    /** @type {?} */
    GaEventDirective.prototype.gaLabel;
    /** @type {?} */
    GaEventDirective.prototype.label;
    /** @type {?} */
    GaEventDirective.prototype.gaValue;
    /** @type {?} */
    GaEventDirective.prototype.gaInteraction;
    /** @type {?} */
    GaEventDirective.prototype.gaBind;
    /**
     * @type {?}
     * @protected
     */
    GaEventDirective.prototype.$gaCategoryDirective;
    /**
     * @type {?}
     * @protected
     */
    GaEventDirective.prototype.$gaService;
}
